### PR TITLE
Factspage, My facts page, Loader

### PR DESCRIPTION
**Pull request by VasylVynn**
Added Loader to the Facts page. It is showing until the facts are fetched and displayed.
Created "Add to favourite" button to each card of Facts page. This button adds fact to the My facts page. Once the button is clicked it became inactive if the fact has been added.

Created My Facts page and Remove button to each fact that will remove it from this page

### CODE REVIEW

File: src/App.tsx [hunk 1]
```diff
@@ -21,6 +24,7 @@ const App: React.FC = () => {
           <Route path="/login" element={<Login />} />
           <Route path="/register" element={<Registration />} />
           <Route path="/facts" element={<Facts />} />
+          <Route path="/fav_facts" element={<FavFacts />} />
```
Comment (boandriy) [inline src/App.tsx:27]: Please take a look at the React router Dom and take a look at prop named exact. Think If you need it or no. Also. I would suggest adding a 'NOT FOUND 404' page. If someone will enter a url that is not in the list.

Comment (boandriy) [reply src/App.tsx:27]: Also. Although it is useless here - I would suggest you create a separate file that will contain an constant array of objects called routes. Each object should have for example - path, component, exact, etc. And here in the App.tsx use this array to map routes.

File: src/components/Card/styles.ts [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+ 
+
+
+&:hover {
+    box-shadow: 0px 5px 10px 6px rgba(14,115,11,0.38);}
+  
+ 
+
+
+
```
Comment (boandriy) [inline src/components/Card/styles.ts:69]: this space is not needed here.

File: src/components/Card/styles.ts [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+export const CardText = styled.div`
+    text-align: center;
+    font-family: sans-serif;
+  `;
+
+
+
+export const FavButton  = styled.button`
+    
+    background: linear-gradient(to bottom right, #71ef47, #30b34a);
```
Comment (boandriy) [inline src/components/Card/styles.ts:38]: wierd space

File: src/components/Loader/Loader.ts [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+    right: 25%;
+    top: 25%;
+    bottom: 25%;
+    margin: auto;
+    width: 32px;
+    height: 32px;
+    transition: opacity 200ms;
+    animation: ${rotate360} 1s linear;
+    animation-iteration-count: infinite;
+    transition-delay: '200ms' ;
```
Comment (boandriy) [inline src/components/Loader/Loader.ts:29]: Same. Minor - the space before ; is not needed here

File: src/pages/Facts/facts.tsx [hunk 5]
```diff
@@ -11,30 +14,28 @@ const Facts: React.FC = () => {
...
 
-    console.log(facts);
+    const loading = useSelector(selectLoading)
 
-    return <div>
-        <h2>Facts</h2>
-        <button  >fact</button>
-    </div>;
-}
+    const stringFavFacts = favFacts.map(fact => fact.fact)
```
Comment (boandriy) [inline src/pages/Facts/facts.tsx:21]: I would still add type here `stringFavFacts: string[]`

File: src/pages/Facts/facts.tsx [hunk 6]
```diff
@@ -11,30 +14,28 @@ const Facts: React.FC = () => {
...
-
-const mapDispatchToProps = (dispatch: any) => {
-    return {
-        fetchFacts: () => dispatch(fetchFacts())
+    useEffect(() => {
+        dispatch(fetchFacts())
+    }, [])
 
-    }
+    const FactsList = () => (
```
Comment (boandriy) [inline src/pages/Facts/facts.tsx:27]: Try to investigate and add a return type here

File: src/pages/FavFacts/FavFacts.tsx [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+    const favFacts = useSelector(selectFavFacts);
+
+
+
+    const FavFactsList = () => (
+        favFacts.map((favfacts, index) =>
+            <Card key={index}>
+                <CardText >{favfacts.fact}</CardText>
+                <FavButton onClick={() => dispatch(removeFavourite(favfacts.fact))}>Remove</FavButton>
+            </Card>
```
Comment (boandriy) [inline src/pages/FavFacts/FavFacts.tsx:23]: Looks very similar to FactsList I would create a component for this and reuse it in those 2 places

File: src/redux/facts/factReducer.ts [hunk 8]
```diff
@@ -1,40 +1,61 @@
-import { FETCH_FACTS_FAILURE, FETCH_FACTS_REQUEST, FETCH_FACTS_SUCCESS } from "./factsTypes"
+import { ADD_FAV, FETCH_FACTS_FAILURE, FETCH_FACTS_REQUEST, FETCH_FACTS_SUCCESS, REMOVE_FAV } from "./factsTypes"
 
 export interface State  {
     loading:boolean;
-    facts: {}[];
+    facts: Array<{ fact: string, length: number }>;
```
Comment (boandriy) [inline src/redux/facts/factReducer.ts:5]: I would create an interface called Fact something like 
```
interface Fact {
  fact: string;
  length: number;
}
```
And then - here would use `Fact[]`

File: src/redux/facts/factsActions.ts [hunk 9]
```diff
@@ -1,17 +1,16 @@
...
-import {State} from "../../redux/facts/factReducer"
 
 export const fetchFactsRequest = () =>( {
     
         type: FETCH_FACTS_REQUEST
     
 })
 
-export const fetchFactsSuccess = (users:{}[]) =>( {
+export const fetchFactsSuccess = (facts:{}[]) =>( {
```
Comment (boandriy) [inline src/redux/facts/factsActions.ts:10]: Add proper type here

File: src/redux/facts/factsActions.ts [hunk 10]
```diff
@@ -22,13 +21,21 @@ export const fetchFactsFailure = (error:string) =>( {
...
+    payload: key
+})
 
 
 export const fetchFacts = () => {
     return (dispatch:any) =>{
-        dispatch(fetchFactsRequest)
-        axios.get('https://catfact.ninja/facts')
+        dispatch(fetchFactsRequest());
+        axios.get('https://catfact.ninja/facts?page=1&max_length=150&limit=20')
```
Comment (boandriy) [inline src/redux/facts/factsActions.ts:38]: Move it somewhere outside to place where you keep constants and name this part as BASE_URL -https://catfact.ninja/
Then the request will be : BASE_URL+'facts...'etc.

File: src/redux/facts/factsActions.ts (hunk 9 above)
Comment (VasylVynn) [reply src/redux/facts/factsActions.ts:10]: facts:{fact:String;length:String}[]

File: src/redux/facts/factReducer.ts (hunk 8 above)
Comment (VasylVynn) [reply src/redux/facts/factReducer.ts:5]: Added interface

File: src/pages/FavFacts/FavFacts.tsx (hunk 7 above)
Comment (VasylVynn) [reply src/pages/FavFacts/FavFacts.tsx:23]: Created component FactCard

File: src/redux/facts/factsActions.ts (hunk 10 above)
Comment (VasylVynn) [reply src/redux/facts/factsActions.ts:38]: Moved export const BASE_URL = 'https://catfact.ninja/' to factsTypes.ts

File: src/App.tsx (hunk 1 above)
Comment (VasylVynn) [reply src/App.tsx:27]: Created Routes.tsx component and added map

**Review by boandriy — APPROVED** [review]

**Pull request merged**